### ISSUE TITLE
SoCo raising exception when playing from YouTube Music|

### ISSUE THREAD

**Issue by Barry4679** [SoCo/SoCo#792]
I am subscribed to SoCo AVT events. 
If my Sonos is playing from YouTubeMusic it generates the AVT event OK, but logs the following exception to the log for each track played.

01-20 23:28:15 158.0095 - WARNING soco.events_base:  110 parse_event_xml; Event contains illegal metadatafor 'enqueued_transport_uri_meta_data'.
Error message: 'UPnP class is incorrect. Expected 'object.container.album.musicAlbum.ALBUM', got 'object.container.album.musicAlbum''
The result will be a SoCoFault.

I added a message to display the data causing  the problem:
01-20 23:28:15 159.0295 - WARNING soco.events_base:  118 parse_event_xml; data value was <DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" xmlns:r="urn:schemas-rinconnetworks-com:metadata-1-0/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/"><item id="1004004cALkSOiGdMl9FQFJGE1Ic8S2rmLFHkNfJ69CFi-KR6SopsRn2" parentID="10fe1044ALkSOiEkIOVrL0HH1ihrgR_Gj1GTz0ZqOUgv4vwS1aNttlQt6jwZTkbBPq7CgzFt3E_p20LznoFNX_xmBB3JTSKcZIve9E98FBLjXTA4m6U3xXATU1i3H_DU6ym61uEu" restricted="true"><dc:title>Schnittke: Tango (Arr. by Andriy Rakhmanin for Violin and Piano) (From &quot;Agony&quot;)</dc:title><upnp:class>object.container.album.musicAlbum.#ALBUM</upnp:class><upnp:album>Schnittke: Tango (Arr. by Andriy Rakhmanin for Violin and Piano) (From &quot;Agony&quot;)</upnp:album><dc:creator>Daniel Hope, Alexey Botvinov</dc:creator><upnp:albumArtURI>https://lh3.googleusercontent.com/ictx7j6cS24dlX-MTtEYFoHrORicbEabuFfpgSS4Le9P8g0mnrTukCCNVeK7zF-lHAz
CRSRQiQ-nXQ7J=w544-h544-l90-rj</upnp:albumArtURI><r:description>New releases</r:description><desc id="cdudn" nameSpace="urn:schemas-rinconnetworks-com:metadata-1-0/">SA_RINCON72711_X_#Svc72711-0-Token</desc></item></DIDL-Lite>

I see the problem with versions 20 & 21 of SoCo.

The issue is not corrupting any part of the AVT event that I care about, but I would really like to stop soco.events_base from filling my app log.

**Comment by pwt** [comment]
SoCo's default logging level is `WARN`. To suppress WARN level logging and below, use:

```
import logging
logging.disable(logging.WARN)
```

There are a number of warnings like the one you've encountered, for types that aren't understood.

**Comment by pwt** [comment]
> Your suggestion would also suppress all info & warning logging from my app wouldn't it?

Yes, just trying to keep it simple, but you obviously know your way around logging already!

It's quite possible that some SoCo logging is at the wrong severity; happy to take suggestions for changes. The warning in question looks like it's correctly classified, however.

Does your approach of reaching in to the `events_base` module work for requests and urllib3?

Alternatively, you might want selectively to alter the logging level in some places. Haven't tried this, but it looks like a handy way of doing it:

https://docs.python.org/3/howto/logging-cookbook.html#using-a-context-manager-for-selective-logging.

**Comment by Barry4679** [comment]
> Does your approach of reaching in to the events_base module work for requests and urllib3?

Yes it does. I had to add 12 of them to get SoCo quietened enough. ... eg. urllib3.connectionpool... It is something you might think of, because in a long running app it is important to see unexpected failures, but too much traffic swamps data from the app's own log. It is very bad at the info level during LAN instability .IIRC some of the messages included lengthy exceptions with stack logs ... although my app also runs Flask, which was also contributing to the problem.

It gets worse when I set my app into debug mode. The amount of SoCo debug messages made it very difficult.

> It's quite possible that some SoCo logging is at the wrong severity; happy to take suggestions for changes. The warning in question looks like it's correctly classified, however.

Some way that the client app could de-elevate or disable all SoCo logging, on the fly, would be really useful. Also it would be good if SoCo had option to quieten, or summarise, chatter from the sub-contractors like urllib3 & requests,

What is the status of the areas of soco that use the soco.ms_data_structures module? I would need to check again, but I think that there are some documentation errors. And that area seems be be somewhat a work in progress. 

Thanks for the link to the logging recipe. I may have seen it before , but I didn't appreciate its elegance until you draw it my attention.

It would not have been suitable for what I needed in this instance. My app can run on a headless device, ie. a Raspberry Pi, or on a NAS. And it has multiple threads, ie:
* one monitoring the LAN for Sonos units dropping in or out of view
* another receiving AVT and ZGT events
* another running Flask so the app may be controlled via html commands from a PC
* etc

One of the html commands puts the app into debug mode, so I needed something that affected the whole app, including all threads, ie. not just code within an enclosure.

**Comment by pwt** [comment]
> Some way that the client app could de-elevate or disable all SoCo logging, on the fly, would be really useful. Also it would be good if SoCo had option to quieten, or summarise, chatter from the sub-contractors like urllib3 & requests

For SoCo at least, I seem to be able to get away with using `logging.getLogger().setLevel(logging.CRITICAL)` (etc.) to change the logging level on the fly.

 > What is the status of the areas of soco that use the soco.ms_data_structures module? I would need to check again, but I think that there are some documentation errors. And that area seems be be somewhat a work in progress.

Sorry, that's not an area of the code I can claim any expertise in at the moment. I know that Kenneth is doing lots of work on reinstating music service support, but that it's not simple.

**Comment by amelchio** [comment]
I agree that SoCo is noisy and for example logs at high log levels for cases that the client code can handle or in situations where a music service misbehaves. Having to set the log level explicitly is inconvenient and one risks missing a log line that was actually important.
